nibabel/streamlines/array_sequence.py

```diff
@@ -0,0 +1,267 @@
+        """
+        if len(elements) == 0:
+            return
+
+        if self._data.ndim == 0:
+            elem = np.asarray(elements[0])
+            self._data = np.zeros((0, elem.shape[1]), dtype=elem.dtype)
+
+        next_offset = self._data.shape[0]
+
+        if is_array_sequence(elements):
+            self._data.resize((self._data.shape[0] + sum(elements._lengths),
+                               self._data.shape[1]))
+
+            offsets = []
+            for offset, length in zip(elements._offsets, elements._lengths):
+                offsets.append(next_offset)
+                chunk = elements._data[offset:offset + length]
+                self._data[next_offset:next_offset + length] = chunk
+                next_offset += length
+
+            self._lengths = np.r_[self._lengths, elements._lengths]
+            self._offsets = np.r_[self._offsets, offsets]
+
+        else:
+            self._data = np.concatenate([self._data] + list(elements), axis=0)
+            lengths = list(map(len, elements))
+            self._lengths = np.r_[self._lengths, lengths]
+            self._offsets = np.r_[self._offsets,
+                                  np.cumsum([next_offset] + lengths)[:-1]]
+
+    def copy(self):
+        """ Creates a copy of this :class:`ArraySequence` object. """
+        # We do not simply deepcopy this object since we might have a chance
+        # to use less memory. For example, if the array sequence being copied
+        # is the result of a slicing operation on a array sequence.
+        seq = ArraySequence()
+        total_lengths = np.sum(self._lengths)
+        seq._data = np.empty((total_lengths,) + self._data.shape[1:],
+                             dtype=self._data.dtype)
+
+        next_offset = 0
+        offsets = []
+        for offset, length in zip(self._offsets, self._lengths):
+            offsets.append(next_offset)
+            chunk = self._data[offset:offset + length]
+            seq._data[next_offset:next_offset + length] = chunk
+            next_offset += length
+
+        seq._offsets = np.asarray(offsets)
+        seq._lengths = self._lengths.copy()
+
+        return seq
+
+    def __getitem__(self, idx):
+        """ Gets sequence(s) through advanced indexing.
+
+        Parameters
+        ----------
+        idx : int or slice or list or ndarray
+            If int, index of the element to retrieve.
+            If slice, use slicing to retrieve elements.
+            If list, indices of the elements to retrieve.
+            If ndarray with dtype int, indices of the elements to retrieve.
+            If ndarray with dtype bool, only retrieve selected elements.
+
+        Returns
+        -------
+        ndarray or :class:`ArraySequence`
+            If `idx` is an int, returns the selected sequence.
+            Otherwise, returns a :class:`ArraySequence` object which is view
+            of the selected sequences.
+        """
+        if isinstance(idx, (int, np.integer)):
```

Comment:
This check will fail for Python Long types on Python 2 - and they appear relatively often:

```
In [5]: isinstance(1L, (int, np.integer))
Out[5]: False
```

Better check is:

```
In [9]: import numbers

In [10]: isinstance(1L, numbers.Integral)
Out[10]: True

In [11]: isinstance(1, numbers.Integral)
Out[11]: True

In [12]: isinstance(np.int(1), numbers.Integral)
Out[12]: True

In [13]: isinstance(np.int32(1), numbers.Integral)
Out[13]: True
```

And, in fact, for best back-compatibility with numpy (back to 1.6):

```
isinstance(1L, (numbers.Integral, np.integer))
```

because this will catch the numpy integer types as well, before they were added to the `numbers.Integral` collection.